<a href="https://colab.research.google.com/github/shaheerzubery/NLP-/blob/main/Sentence_similarity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Sentence transformer usng bert


So, Today we are trying to check the sementic and similarity  between sentence using pre trained model Sbert

In [1]:
!pip install -Uqqq pip --progress-bar off
!pip install -qqq sentence-transformers==2.2.2 --progress-bar off

In [2]:
from sentence_transformers import SentenceTransformer, util

Now create a model

In [3]:
model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')

I'll provide the documentation in the description below, from where i used this pre trained model 

In [4]:
corpus = [
    'weather to be decrease to -1 degree by today',
    'it is a cold weather',
    'you can play football today',
    'you have to wear hot clothes to keep yourself warm',]

In [5]:
model.max_seq_length

384

This shws that the max length of character that are sufficent for this pre trained weights are 384

Now we will train our corpus on the pre trained model

In [6]:
corpus_embedding = model.encode(corpus,show_progress_bar=True, convert_to_tensor=True)

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

In [14]:
query = "which clothes to wear today"

In [15]:
query_embedding = model.encode(query,convert_to_tensor=True)

In [16]:
corpus_embedding.shape

torch.Size([4, 768])

In [17]:
util.semantic_search(query_embedding,corpus_embedding[0])

[[{'corpus_id': 0, 'score': 0.23167961835861206}]]

This show that this sentence is similar to the query one

In [18]:
result = util.semantic_search(query_embedding,corpus_embedding)[0]

In [19]:
for item in result:
  print(round(item['score'],2),corpus[item['corpus_id']])

0.32 you have to wear hot clothes to keep yourself warm
0.3 you can play football today
0.23 weather to be decrease to -1 degree by today
0.17 it is a cold weather


This shows that which sentence is more similar to the query, 

This is based on pretrained model you can create your own model as well,

how to create it will show you next video